Sc 3784/databricks download file from dbfs

Shamwayne opened the pull request.

Shamwayne committed:
init download file logic

Shamwayne committed:
Update download_file_from_dbfs.py

Shamwayne committed:
Update download_file_from_dbfs.py

Shamwayne committed:
Update download_file_from_dbfs.py

Shamwayne committed:
added error codes

Shamwayne committed:
Update download_file_from_dbfs.py

Shamwayne committed:
Merge branch 'main' into sc-3784/databricks-download-file-from-dbfs

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,80 @@
+                "length": 1024 # 1MB of data 
+            }
+            read_response = client.get(read_handle_endpoint, params=payload)
+            if read_response.status_codes == requests.codes.ok:
```
Errors on `status_codes`. Believe it needs to be `status_code`

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,80 @@
+    source_file_name = args.source_file_name
+    source_folder_name = args.source_folder_name
+    dest_file_name = args.dest_file_name
+    dest_folder_name = args.dest_folder_name
```
Needs to be cleaned using shipyard_utils to ensure that it doesn't matter if there are preceding/trailing slashes.

Also needs to create directory structure if it doesn't already exist.

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,80 @@
+    # create client
+    client = helpers.DatabricksClient(access_token, instance_id)
+    # create file paths
+    if not source_folder_name:
```
Needs to be cleaned with shipyard_utils so the placement of `/` doesn't impact anything.

blakeburch reviewed (commented):
1) This doesn't download the files correctly. It just stores the data in what looks like an encoded format. The file should be downloaded and openable in whatever native file format it was originally in.

A sample file with a 3 columns, a header row, and 5 rows of data returned `77u/Rmlyc3QgTmFtZSxMYXN0IE5hbWUsTnVtYmVyDQpBbGJlcnQsU21pdGgsMTIzDQpCb2IsR25vbCwyMzQNCkNoYXJsZXMsU2luYSwzNDUNCkRhcnlsLERlYnVzc3ksNDU2DQpFdWd1bmUsSG9ybnNiZXJnLDU2Nw==`
2) Apologies if it wasn't explicity, but this needs to be able to download multiple files from DBFS by including a `source-file-name-match-type` field of regex. Shipyard utils includes the functionality to manage that correctly.
3) Specific errors need to be put in place for things like wrong access-token, wrong instance-id, file doesn't exist on databricks, etc.

Shamwayne committed:
added base64 conversion

Shamwayne committed:
Update download_file_from_dbfs.py

blakeburch commented:
@Shamwayne Small files seem to download fine. Medium files (10mb) always run into `Connection Error: Failed to reach server at /dbfs/read` before completing. I confirmed that only 22kb/9mb actually got downloaded.

Also, the other comments on this PR still need to be addressed.

Shamwayne committed:
Fixed download rate and added HTTP Keep-alive

- Fixed download speed from 1KB per request to 1MB per request
- Added HTTP connection reuse for the same requests made using Databricks Client

Shamwayne committed:
Update download_file_from_dbfs.py

re-added combine folder path logic

Shamwayne committed:
Added error handling and other fixes

Added error handling for Databricks as well as initial handling for regex issues

Shamwayne committed:
added list DBFS files functionality

added the ability to list all files within a DBFS folder for use in regex matching

Shamwayne committed:
Update helpers.py

Shamwayne committed:
Update download_file_from_dbfs.py

added regex matching

Shamwayne committed:
added checks for file_types

Shamwayne committed:
update to latest helpers.py

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,132 @@
+    # otherwise download the file normally
+    else:
+        source_file_path = shipyard.files.combine_folder_and_file_name(
+                        source_folder_name, file_name)
```
local variable 'file_name' referenced before assignment. Pretty sure this needs to be `source_file_name`

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,132 @@
+        source_file_path = shipyard.files.combine_folder_and_file_name(
+                        source_folder_name, file_name)
+        destination_file_path = shipyard.files.combine_folder_and_file_name(
+                            dest_folder_name, dest_file_name)
```
Needs to use shipyard_utils to create the folder structure. Right now, if a folder is provided, it just returns ` No such file or directory:`

blakeburch replied:
@Shamwayne This is still occuring. If I provide any dest-folder-name that doesn't already exist, it fails.

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,132 @@
+        
+    # get list of all potential file matches
+    if source_file_name_match_type == 'regex_match':
+        files = helpers.list_dbfs_files(client, source_folder_name)
```
When 0 matches are found, it still says "downloading" and results in an exit code of 0. Need to error handle for that.

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,132 @@
+        dest_folder_name = os.getcwd()
+        
+    # get list of all potential file matches
+    if source_file_name_match_type == 'regex_match':
```
No need to print the matching file names if you upgrade to use shipyard_utils==0.1.3 (changed as of today)

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,132 @@
+    dest_folder_name = args.dest_folder_name
+    source_file_name_match_type = args.source_file_name_match_type
+    # create client
+    client = helpers.DatabricksClient(access_token, instance_id)
```
Still need to error handle somewhere for an incorrect instance id. Currently just returns:

Error sending request to /dbfs/list: HTTPSConnectionPool(host='randomid.cloud.databricks.comm', port=443): Max retries exceeded with url: /api/2.0/dbfs/list?path=%2FFileStore%2F (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x1043990a0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,132 @@
+    client = helpers.DatabricksClient(access_token, instance_id)
+    
+    if not source_folder_name:
+        source_folder_name = '/FileStore/'
```
Probably make more sense for this to be a default of the argument?

blakeburch reviewed (changes requested):
Minor fixes needed. @ericelsken - you're good to go ahead and code review since all of the core functionality is here now.

Shamwayne committed:
fix filename

Shamwayne committed:
Update download_file_from_dbfs.py

Shamwayne committed:
add check for zero matches

ericelsken commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,134 @@
+                read_response = client.get(read_handle_endpoint, params=payload)
+            except Exception:
+                print(f'Connection Error for {read_handle_endpoint}')
+                print(f"Data downloaded: {offset/1024}MB")
```
Why are you dividing by 1024 and (1024*1024) or just printing the offset directly?

Shamwayne replied:
Do you suggest i just leave the downloaded data in terms of bytes or should i remove that information completely from the output?

ericelsken replied:
I think leaving it in bytes is fine. [](edited-with-Axolo)

ericelsken commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,134 @@
+                else:
+                    file.write(data.decode('utf-8'))
+                # check if there is still more data left to read
+                if response_data['bytes_read'] == ONE_MB:
+                    offset += ONE_MB
+                else: # that was the last of the data in the file
+                    break
```
Is there no other way to know when the download is done? 
It seems weird that you have to check the response size against the request size to determine whether or not to continue. 
What if the response is less than ONE_MB - the service returns less than requested? Is the service guaranteeing to return the exact number of bytes requested, and does that play nicely with the base64 encoding? 
Can you link to docs for this endpoint?  [](edited-with-Axolo)

Shamwayne replied:
Documentation for reads comes here: https://docs.databricks.com/dev-tools/api/latest/dbfs.html#read
Most of the download logic itself was derived from how file uploads are handled though which also deals with base64 data: https://docs.databricks.com/dev-tools/api/latest/examples.html#upload-a-big-file-into-dbfs

ericelsken replied:
Okay, so it looks like you can use the response size being less than the requested size to determine if you can stop reading.
Can you test / provide an example of what happens when the `offset` equals the file size, and `length` is some positive value?

I think we should update the logic to introduce a `length` or `requested_read_length` parameter, use that value in the params, and then compare the resulting length to that value. If the resulting read is less than the request, then stop reading. Otherwise, continue. But, this depends on my question / example above. [](edited-with-Axolo)

Shamwayne replied:
If i'm getting your example: you're asking say the offset is some value like say 10024 and the file size itself is 10024? That means the amount of data downloaded is equivalent to the amount of data the file has so even if another call is performed the next call will just return 0 bytes (no more data left to download) and close the file stream. 

> I think we should update the logic to introduce a `length` or `requested_read_length` parameter, use that value in the params, and then compare the resulting length to that value. If the resulting read is less than the request, then stop reading. Otherwise, continue. But, this depends on my question / example above.

logic of the code already works somewhat similarly to what you're suggesting unless you're requesting for the ability to perform partial downloads?

ericelsken replied:
Yes, your example is correct. I'm not asking for partial downloads, just that the example case is accounted for and works correctly.

ericelsken commented on databricks_blueprints/errors.py:
```diff
@@ -3,6 +3,14 @@
 
 EXIT_CODE_INVALID_CREDENTIALS = 200
 EXIT_CODE_INVALID_INSTANCE = 201
+EXIT_CODE_BAD_REQUEST = 202
```
This is a copy of EXIT_CODE_INVALID_CLUSTER. 
You also have multiple names for the 0 value. You can create just one constant and use it in multiple places indicating success.  [](edited-with-Axolo)

ericelsken commented on databricks_blueprints/helpers.py:
```diff
@@ -1,5 +1,6 @@
 import sys
 import requests
+from pathlib import Path
```
Can we rename this file to client? Generally the name helpers is too ... general. And the point of the module is to expose our DatabricksClient.

Shamwayne committed:
Update download_file_from_dbfs.py

clean the folder name

Shamwayne committed:
Update requirements.txt

Shamwayne committed:
rename to databricks client

Shamwayne committed:
fix error codes

Shamwayne committed:
typo

blakeburch reviewed (commented):
@shamwayne I can't effectively test this because it keeps returning `AttributeError: 'DatabricksClient' object has no attribute 'list_dbfs_files'`

Shamwayne committed:
Update download_file_from_dbfs.py

Shamwayne commented:
retry it right now, there was one commit missing from where i'm testing from (Linux based for the file paths) vs my Windows environment which is why

Shamwayne committed:
Update databricks_client.py

Shamwayne committed:
download file fixes and regex handling

Shamwayne commented:
@blakeburch please review and test this blueprint

Shamwayne committed:
Update databricks_client.py

blakeburch commented on databricks_blueprints/download_file_from_dbfs.py:
```diff
@@ -0,0 +1,139 @@
+            source_file_path = file_name
+            # create destination file name
+            dest_file_name = shipyard.files.extract_file_name_from_source_full_path(file_name)
+            dest_file_name = shipyard.files.enumerate_destination_file_name(dest_file_name, index)
```
enumeration should only happen if:
- dest-file-name is provided AND
- more than one file is matched to

See - <https://www.shipyardapp.com/docs/reference/blueprints/blueprint-library/match-type/#effects-on-destination-1> [](edited-with-Axolo)

blakeburch reviewed (commented):
@shamwayne Main commentary:

- Regex match looks to be solid. Woohoo!
- Enumeration logic is off. See the documentation again for specifications.
- Files still can't be downloaded if the local folder doesn't exist.

Shamwayne committed:
changed destination file rules

Shamwayne committed:
better databricks logic

Shamwayne committed:
Update download_file_from_dbfs.py

Commit:
Minor edits for enumeration and defaults

blakeburch reviewed (approved):
Made minor edits. Otherwise good to go.

blakeburch commented:
@ericelsken Did you want to take a 2nd look at this one? Or are you good?
